File: rust/src/issue_update.rs
//! Issue update workflow.

use chrono::Utc;
use std::fs;
use std::path::Path;

use crate::agent_metadata::assign_agent_metadata_if_incomplete;
use crate::config_loader::load_project_configuration;
use crate::error::KanbusError;
use crate::event_history::{build_update_events, now_timestamp};
use crate::file_io::get_configuration_path;
use crate::issue_creation::resolve_issue_identifier;
use crate::issue_files::read_issue_from_file;
use crate::issue_lookup::load_issue_from_project;
use crate::issue_mutation::{persist_issue_mutation, PersistIssueMutationRequest};
use crate::models::{AgentMetadata, IssueData};
use crate::users::get_current_user;
use crate::workflows::{
    apply_transition_side_effects, validate_status_transition, validate_status_value,
};

/// Result of an issue update operation.
#[derive(Debug, Clone)]
pub struct IssueUpdateResult {
    /// Updated issue data.
    pub issue: IssueData,
    /// Whether disk state changed.
    pub changed: bool,
}

/// Update an issue and persist it to disk.
///
/// # Arguments
/// * `root` - Repository root path.
/// * `identifier` - Issue identifier.
/// * `title` - Updated title if provided.
/// * `description` - Updated description if provided.
/// * `status` - Updated status if provided.
/// * `assignee` - Updated assignee if provided.
/// * `claim` - Whether to claim the issue.
///
/// # Errors
/// Returns `KanbusError` if the update fails.
#[allow(clippy::too_many_arguments)]
pub fn update_issue(
    root: &Path,
    identifier: &str,
    title: Option<&str>,
    description: Option<&str>,
    status: Option<&str>,
    assignee: Option<&str>,
    priority: Option<u8>,
    claim: bool,
    validate: bool,
    add_labels: &[String],
    remove_labels: &[String],
    set_labels: Option<&str>,
    parent: Option<&str>,
    issue_type: Option<&str>,
    agent: Option<AgentMetadata>,
) -> Result<IssueUpdateResult, KanbusError> {
    let fields_requested = title.is_some()
        || description.is_some()
        || status.is_some()
        || claim
        || assignee.is_some()
        || priority.is_some()
        || !add_labels.is_empty()
        || !remove_labels.is_empty()
        || set_labels.is_some()
        || parent.is_some()
        || issue_type.is_some()
        || agent.is_some();

    let lookup = load_issue_from_project(root, identifier)?;
    let before_issue = lookup.issue.clone();
    let config_path = get_configuration_path(lookup.project_dir.as_path())?;
    let configuration = load_project_configuration(&config_path)?;

    let mut updated_issue = lookup.issue.clone();
    let current_time = Utc::now();

    let mut resolved_status = if claim { Some("in_progress") } else { status };
    let mut resolved_type = issue_type.map(str::trim).filter(|value| !value.is_empty());
    if resolved_type == Some(updated_issue.issue_type.as_str()) {
        resolved_type = None;
    }

    let mut updated_title: Option<String> = None;
    if let Some(new_title) = title {
        let normalized_title = new_title.trim();
        if normalized_title.to_lowercase() != updated_issue.title.trim().to_lowercase() {
            if let Some(duplicate_identifier) = find_duplicate_title(
                &lookup.project_dir.join("issues"),
                normalized_title,
                &updated_issue.identifier,
            )? {
                return Err(KanbusError::IssueOperation(format!(
                    "duplicate title: \"{}\" already exists as {}",
                    normalized_title, duplicate_identifier
                )));
            }
            updated_title = Some(normalized_title.to_string());
        }
    }

    let mut updated_description: Option<String> = None;
    if let Some(new_description) = description {
        let normalized_description = new_description.trim();
        if normalized_description != updated_issue.description {
            updated_description = Some(normalized_description.to_string());
        }
    }

    let mut updated_assignee: Option<String> = None;
    if let Some(new_assignee) = assignee {
        if updated_issue.assignee.as_deref() != Some(new_assignee) {
            updated_assignee = Some(new_assignee.to_string());
        }
    }

    let mut updated_priority: Option<i32> = None;
    if let Some(new_priority) = priority {
        if validate && !configuration.priorities.contains_key(&new_priority) {
            return Err(KanbusError::IssueOperation("invalid priority".to_string()));
        }
        if updated_issue.priority != new_priority as i32 {
            updated_priority = Some(new_priority as i32);
        }
    }

    if resolved_status.is_some() && resolved_status == Some(updated_issue.status.as_str()) {
        resolved_status = None;
    }

    let mut updated_labels: Option<Vec<String>> = None;
    if set_labels.is_some() || !add_labels.is_empty() || !remove_labels.is_empty() {
        let mut labels = if let Some(value) = set_labels {
            value
                .split(',')
                .map(|label| label.trim().to_string())
                .filter(|l| !l.is_empty())
                .collect::<Vec<_>>()
        } else {
            updated_issue.labels.clone()
        };
        for label in add_labels {
            let trimmed = label.trim();
            if !trimmed.is_empty() && !labels.iter().any(|l| l.eq_ignore_ascii_case(trimmed)) {
                labels.push(trimmed.to_string());
            }
        }
        if !remove_labels.is_empty() {
            labels.retain(|label| {
                !remove_labels
                    .iter()
                    .any(|r| label.eq_ignore_ascii_case(r.trim()))
            })
        }
        if labels != updated_issue.labels {
            updated_labels = Some(labels);
        }
    }

    let mut updated_parent: Option<String> = None;
    if let Some(parent_candidate) = parent {
        let issues_dir = lookup.project_dir.join("issues");
        let resolved_parent =
            resolve_issue_identifier(&issues_dir, &configuration.project_key, parent_candidate)?;
        if updated_issue.parent.as_deref() != Some(resolved_parent.as_str()) {
            if validate {
                let parent_path = issues_dir.join(format!("{resolved_parent}.json"));
                if !parent_path.exists() {
                    return Err(KanbusError::IssueOperation("not found".to_string()));
                }
                let parent_issue = read_issue_from_file(&parent_path)?;
                crate::hierarchy::validate_parent_child_relationship(
                    &configuration,
                    &parent_issue.issue_type,
                    resolved_type.unwrap_or(&updated_issue.issue_type),
                )?;
            }
            updated_parent = Some(resolved_parent);
        }
    }

    if validate {
        if let Some(new_type) = resolved_type {
            let is_known = configuration
                .hierarchy
                .iter()
                .chain(configuration.types.iter())
                .any(|entry| entry == new_type);
            if !is_known {
                return Err(KanbusError::IssueOperation(
                    "unknown issue type".to_string(),
                ));
            }

            let issues_dir = lookup.project_dir.join("issues");

            if let Some(parent_identifier) = updated_issue.parent.as_deref() {
                let parent_path = issues_dir.join(format!("{parent_identifier}.json"));
                if !parent_path.exists() {
                    return Err(KanbusError::IssueOperation("not found".to_string()));
                }
                let parent_issue = read_issue_from_file(&parent_path)?;
                crate::hierarchy::validate_parent_child_relationship(
                    &configuration,
                    &parent_issue.issue_type,
                    new_type,
                )?;
            }

            for entry in
                fs::read_dir(&issues_dir).map_err(|error| KanbusError::Io(error.to_string()))?
            {
                let entry = entry.map_err(|error| KanbusError::Io(error.to_string()))?;
                let child_path = entry.path();
                if child_path.extension().and_then(|ext| ext.to_str()) != Some("json") {
                    continue;
                }
                let child_issue = read_issue_from_file(&child_path)?;
                if child_issue.parent.as_deref() != Some(updated_issue.identifier.as_str()) {
                    continue;
                }
                crate::hierarchy::validate_parent_child_relationship(
                    &configuration,
                    new_type,
                    &child_issue.issue_type,
                )?;
            }
        }
    }

    let assigned_agent = assign_agent_metadata_if_incomplete(updated_issue.agent.clone(), agent)?;
    let agent_changed = assigned_agent != updated_issue.agent;

    if resolved_status.is_none()
        && resolved_type.is_none()
        && updated_title.is_none()
        && updated_description.is_none()
        && updated_assignee.is_none()
        && updated_priority.is_none()
        && updated_labels.is_none()
        && updated_parent.is_none()
        && !agent_changed
    {
        if fields_requested {
            return Ok(IssueUpdateResult {
                issue: before_issue,
                changed: false,
            });
        }
        return Err(KanbusError::IssueOperation(
            "no updates requested".to_string(),
        ));
    }

    if let Some(new_status) = resolved_status {
        if validate {
            let type_for_validation = resolved_type.unwrap_or(&updated_issue.issue_type);
            validate_status_value(&configuration, type_for_validation, new_status)?;
            validate_status_transition(
                &configuration,
                type_for_validation,
                &updated_issue.status,
                new_status,
            )?;
        }
        updated_issue = apply_transition_side_effects(&updated_issue, new_status, current_time);
        updated_issue.status = new_status.to_string();
    }
    if let Some(new_type) = resolved_type {
        updated_issue.issue_type = new_type.to_string();
    }

    if let Some(new_title) = updated_title {
        updated_issue.title = new_title;
    }
    if let Some(new_description) = updated_description {
        updated_issue.description = new_description;
    }
    if let Some(new_assignee) = updated_assignee {
        updated_issue.assignee = Some(new_assignee);
    }
    if let Some(new_priority) = updated_priority {
        updated_issue.priority = new_priority;
    }
    if let Some(new_labels) = updated_labels {
        updated_issue.labels = new_labels;
    }
    if let Some(new_parent) = updated_parent {
        updated_issue.parent = Some(new_parent);
    }
    if agent_changed {
        updated_issue.agent = assigned_agent;
    }

    let policies_dir = lookup.project_dir.join("policies");
    if policies_dir.is_dir() {
        let policy_documents = crate::policy_loader::load_policies(&policies_dir)?;
        if !policy_documents.is_empty() {
            let issues_dir = lookup.project_dir.join("issues");
            let all_issues = crate::issue_listing::load_issues_from_directory(&issues_dir)?;
            let context = crate::policy_context::PolicyContext {
                current_issue: Some(before_issue.clone()),
                proposed_issue: updated_issue.clone(),
                transition: resolved_status.map(|s| crate::policy_context::StatusTransition {
                    from: before_issue.status.clone(),
                    to: s.to_string(),
                }),
                operation: crate::policy_context::PolicyOperation::Update,
                project_configuration: configuration.clone(),
                all_issues,
            };
            crate::policy_evaluator::evaluate_policies(&context, &policy_documents)?;
        }
    }

    let occurred_at = now_timestamp();
    let actor_id = get_current_user();
    let events = build_update_events(&before_issue, &updated_issue, &actor_id, &occurred_at);
    let result = persist_issue_mutation(&PersistIssueMutationRequest {
        project_dir: lookup.project_dir.clone(),
        issue_path: lookup.issue_path.clone(),
        issue: updated_issue.clone(),
        actor_id,
        events,
        root: root.to_path_buf(),
        before_issue: Some(before_issue),
        relocate_to: None,
        regenerate_right_now: true,
    })?;
    let updated_issue = result.issue;

    if lookup.issue_path.parent() == Some(lookup.project_dir.join("issues").as_path()) {
        let event_id = result.events.first().map(|event| event.event_id.clone());
        crate::gossip::publish_issue_mutation(
            root,
            &lookup.project_dir,
            &updated_issue,
            event_id,
            "issue.mutated",
        );
    }

    Ok(IssueUpdateResult {
        issue: updated_issue,
        changed: true,
    })
}

fn find_duplicate_title(
    issues_dir: &Path,
    title: &str,
    current_identifier: &str,
) -> Result<Option<String>, KanbusError> {
    let normalized_title = title.trim().to_lowercase();
    for entry in fs::read_dir(issues_dir).map_err(|error| KanbusError::Io(error.to_string()))? {
        let entry = entry.map_err(|error| KanbusError::Io(error.to_string()))?;
        let path = entry.path();
        if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        if path
            .file_stem()
            .and_then(|stem| stem.to_str())
            .map(|stem| stem == current_identifier)
            .unwrap_or(false)
        {
            continue;
        }
        let issue = match read_issue_from_file(&path) {
            Ok(issue) => issue,
            Err(_) => continue,
        };
        if issue.title.trim().to_lowercase() == normalized_title {
            return Ok(Some(issue.identifier));
        }
    }
    Ok(None)
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::{TimeZone, Utc};
    use tempfile::TempDir;

    fn issue(identifier: &str, title: &str) -> IssueData {
        let timestamp = Utc.with_ymd_and_hms(2026, 3, 6, 0, 0, 0).unwrap();
        IssueData {
            identifier: identifier.to_string(),
            title: title.to_string(),
            description: String::new(),
            issue_type: "task".to_string(),
            status: "open".to_string(),
            priority: 2,
            assignee: None,
            creator: None,
            parent: None,
            labels: Vec::new(),
            dependencies: Vec::new(),
            comments: Vec::new(),
            created_at: timestamp,
            updated_at: timestamp,
            closed_at: None,
            agent: None,
            right_now_summary: None,
            right_now_updated_at: None,
            custom: std::collections::BTreeMap::new(),
        }
    }

    fn write_issue(issues_dir: &Path, issue: &IssueData) {
        std::fs::write(
            issues_dir.join(format!("{}.json", issue.identifier)),
            serde_json::to_vec(issue).expect("serialize issue"),
        )
        .expect("write issue");
    }

    #[test]
    fn find_duplicate_title_matches_case_insensitively() {
        let temp_dir = TempDir::new().expect("tempdir");
        let issues_dir = temp_dir.path().join("issues");
        std::fs::create_dir_all(&issues_dir).expect("create issues");
        write_issue(&issues_dir, &issue("kanbus-aaa", "Duplicate Title"));
        write_issue(&issues_dir, &issue("kanbus-bbb", "Different"));

        let duplicate =
            find_duplicate_title(&issues_dir, "duplicate title", "kanbus-bbb").expect("lookup");

        assert_eq!(duplicate.as_deref(), Some("kanbus-aaa"));
    }

    #[test]
    fn find_duplicate_title_ignores_current_identifier_and_invalid_json() {
        let temp_dir = TempDir::new().expect("tempdir");
        let issues_dir = temp_dir.path().join("issues");
        std::fs::create_dir_all(&issues_dir).expect("create issues");
        write_issue(&issues_dir, &issue("kanbus-aaa", "Current"));
        std::fs::write(issues_dir.join("broken.json"), "{").expect("write invalid");

        let duplicate = find_duplicate_title(&issues_dir, "Current", "kanbus-aaa").expect("lookup");

        assert!(duplicate.is_none());
    }
}

File: python/src/kanbus/issue_update.py
"""Issue update workflow."""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from dataclasses import dataclass

from pydantic import ValidationError

from kanbus.config_loader import load_project_configuration
from kanbus.issue_mutation import PersistIssueMutationRequest, persist_issue_mutation
from kanbus.issue_files import read_issue_from_file
from kanbus.issue_lookup import (
    IssueLookupError,
    load_issue_from_project,
    resolve_issue_identifier,
)
from kanbus.hierarchy import InvalidHierarchyError, validate_parent_child_relationship
from kanbus.agent_metadata import (
    AgentMetadataResolutionError,
    assign_agent_metadata_if_incomplete,
)
from kanbus.models import AgentMetadata, IssueData
from kanbus.project import get_configuration_path
from kanbus.workflows import (
    InvalidTransitionError,
    apply_transition_side_effects,
    validate_status_transition,
    validate_status_value,
)
from kanbus.event_history import (
    build_update_events,
    now_timestamp,
)
from kanbus.users import get_current_user
from kanbus.gossip import publish_issue_mutation


class IssueUpdateError(RuntimeError):
    """Raised when issue updates fail."""


@dataclass(frozen=True)
class IssueUpdateResult:
    """Result of an issue update operation."""

    issue: IssueData
    changed: bool


def update_issue(
    root: Path,
    identifier: str,
    title: Optional[str],
    description: Optional[str],
    status: Optional[str],
    assignee: Optional[str],
    claim: bool,
    validate: bool = True,
    priority: Optional[int] = None,
    add_labels: Optional[list[str]] = None,
    remove_labels: Optional[list[str]] = None,
    set_labels: Optional[list[str]] = None,
    parent: Optional[str] = None,
    issue_type: Optional[str] = None,
    agent: Optional[AgentMetadata] = None,
) -> IssueUpdateResult:
    """Update an issue and persist it to disk.

    :param root: Repository root path.
    :type root: Path
    :param identifier: Issue identifier.
    :type identifier: str
    :param title: Updated title if provided.
    :type title: Optional[str]
    :param description: Updated description if provided.
    :type description: Optional[str]
    :param status: Updated status if provided.
    :type status: Optional[str]
    :param assignee: Updated assignee if provided.
    :type assignee: Optional[str]
    :param claim: Whether to claim the issue.
    :type claim: bool
    :param priority: Updated priority if provided.
    :type priority: Optional[int]
    :param add_labels: Labels to add.
    :type add_labels: Optional[list[str]]
    :param remove_labels: Labels to remove.
    :type remove_labels: Optional[list[str]]
    :param set_labels: Labels to set (replacing all existing labels).
    :type set_labels: Optional[list[str]]
    :param parent: Updated parent identifier.
    :type parent: Optional[str]
    :param agent: Agent provenance to set when the issue has none or is incomplete.
    :type agent: Optional[AgentMetadata]
    :return: Updated issue data and whether disk state changed.
    :rtype: IssueUpdateResult
    :raises IssueUpdateError: If the update fails.
    """
    fields_requested = (
        title is not None
        or description is not None
        or status is not None
        or claim
        or assignee is not None
        or priority is not None
        or add_labels is not None
        or remove_labels is not None
        or set_labels is not None
        or parent is not None
        or issue_type is not None
        or agent is not None
    )

    try:
        lookup = load_issue_from_project(root, identifier)
    except IssueLookupError as error:
        raise IssueUpdateError(str(error)) from error

    project_dir = lookup.project_dir
    configuration = load_project_configuration(get_configuration_path(project_dir))
    before_issue = lookup.issue
    updated_issue = lookup.issue
    current_time = datetime.now(timezone.utc)

    resolved_status = status
    if claim:
        resolved_status = "in_progress"
    resolved_type = issue_type.strip() if issue_type is not None else None
    if resolved_type == "":
        resolved_type = None

    if resolved_type is not None and resolved_type == updated_issue.issue_type:
        resolved_type = None

    if title is not None:
        normalized_title = title.strip()
        if normalized_title.casefold() == updated_issue.title.strip().casefold():
            title = None
        else:
            duplicate_identifier = _find_duplicate_title(
                project_dir / "issues",
                normalized_title,
                updated_issue.identifier,
            )
            if duplicate_identifier is not None:
                message = (
                    f'duplicate title: "{normalized_title}" '
                    f"already exists as {duplicate_identifier}"
                )
                raise IssueUpdateError(message)
            title = normalized_title

    if description is not None:
        description = description.strip()
        if description == updated_issue.description:
            description = None

    if assignee is not None and assignee == updated_issue.assignee:
        assignee = None

    if resolved_status is not None and resolved_status == updated_issue.status:
        resolved_status = None

    if priority is not None and priority == updated_issue.priority:
        priority = None

    # Handle label operations
    labels = None
    if set_labels is not None:
        if sorted(set_labels) != sorted(updated_issue.labels):
            labels = set_labels
    elif add_labels is not None or remove_labels is not None:
        current_labels = set(updated_issue.labels)
        if add_labels:
            current_labels.update(add_labels)
        if remove_labels:
            current_labels.difference_update(remove_labels)
        candidate_labels = sorted(current_labels)
        if candidate_labels != sorted(updated_issue.labels):
            labels = candidate_labels

    updated_parent: Optional[str] = None
    if parent is not None:
        issues_dir = project_dir / "issues"
        try:
            resolved_parent = resolve_issue_identifier(
                issues_dir, configuration.project_key, parent
            )
        except IssueLookupError as error:
            raise IssueUpdateError(str(error)) from error
        if updated_issue.parent != resolved_parent:
            if validate:
                parent_issue = read_issue_from_file(
                    issues_dir / f"{resolved_parent}.json"
                )
                try:
                    validate_parent_child_relationship(
                        configuration,
                        parent_issue.issue_type,
                        resolved_type or updated_issue.issue_type,
                    )
                except InvalidHierarchyError as error:
                    raise IssueUpdateError(str(error)) from error
            updated_parent = resolved_parent

    if validate and resolved_type is not None:
        valid_types = configuration.hierarchy + configuration.types
        if resolved_type not in valid_types:
            raise IssueUpdateError("unknown issue type")

        issues_dir = project_dir / "issues"

        if updated_issue.parent:
            parent_issue = read_issue_from_file(
                issues_dir / f"{updated_issue.parent}.json"
            )
            try:
                validate_parent_child_relationship(
                    configuration,
                    parent_issue.issue_type,
                    resolved_type,
                )
            except InvalidHierarchyError as error:
                raise IssueUpdateError(str(error)) from error

        for child_path in issues_dir.glob("*.json"):
            child = read_issue_from_file(child_path)
            if child.parent != updated_issue.identifier:
                continue
            try:
                validate_parent_child_relationship(
                    configuration,
                    resolved_type,
                    child.issue_type,
                )
            except InvalidHierarchyError as error:
                raise IssueUpdateError(str(error)) from error

    try:
        assigned_agent = assign_agent_metadata_if_incomplete(updated_issue.agent, agent)
    except AgentMetadataResolutionError as error:
        raise IssueUpdateError(str(error)) from error
    agent_changed = assigned_agent != updated_issue.agent

    if (
        resolved_status is None
        and resolved_type is None
        and title is None
        and description is None
        and assignee is None
        and priority is None
        and labels is None
        and updated_parent is None
        and not agent_changed
    ):
        if fields_requested:
            return IssueUpdateResult(issue=before_issue, changed=False)
        raise IssueUpdateError("no updates requested")

    if resolved_status is not None:
        if validate:
            try:
                validate_status_value(
                    configuration,
                    resolved_type or updated_issue.issue_type,
                    resolved_status,
                )
                validate_status_transition(
                    configuration,
                    resolved_type or updated_issue.issue_type,
                    updated_issue.status,
                    resolved_status,
                )
            except InvalidTransitionError as error:
                raise IssueUpdateError(str(error)) from error
        updated_issue = apply_transition_side_effects(
            updated_issue,
            resolved_status,
            current_time,
        )
        updated_issue = updated_issue.model_copy(update={"status": resolved_status})

    update_fields = {}
    if title is not None:
        update_fields["title"] = title
    if description is not None:
        update_fields["description"] = description
    if assignee is not None:
        update_fields["assignee"] = assignee
    if priority is not None:
        update_fields["priority"] = priority
    if labels is not None:
        update_fields["labels"] = labels
    if updated_parent is not None:
        update_fields["parent"] = updated_parent
    if resolved_type is not None:
        update_fields["issue_type"] = resolved_type
    if agent_changed:
        update_fields["agent"] = assigned_agent

    updated_issue = updated_issue.model_copy(update=update_fields)

    policies_dir = project_dir / "policies"
    if policies_dir.is_dir():
        from kanbus.policy_loader import load_policies
        from kanbus.policy_evaluator import evaluate_policies
        from kanbus.policy_context import (
            PolicyContext,
            PolicyOperation,
            PolicyViolationError,
            StatusTransition,
        )
        from kanbus.issue_listing import load_issues_from_directory

        policy_documents = load_policies(policies_dir)
        if policy_documents:
            issues_dir = project_dir / "issues"
            all_issues = load_issues_from_directory(issues_dir)
            context = PolicyContext(
                current_issue=before_issue,
                proposed_issue=updated_issue,
                transition=(
                    StatusTransition(
                        from_status=before_issue.status,
                        to_status=resolved_status,
                    )
                    if resolved_status
                    else None
                ),
                operation=PolicyOperation.UPDATE,
                project_configuration=configuration,
                all_issues=all_issues,
            )
            try:
                evaluate_policies(context, policy_documents)
            except PolicyViolationError as error:
                raise IssueUpdateError(str(error)) from error

    occurred_at = now_timestamp()
    actor_id = get_current_user()
    events = build_update_events(before_issue, updated_issue, actor_id, occurred_at)
    try:
        result = persist_issue_mutation(
            PersistIssueMutationRequest(
                project_dir=lookup.project_dir,
                issue_path=lookup.issue_path,
                issue=updated_issue,
                actor_id=actor_id,
                events=events,
                before_issue=before_issue,
                root=root,
            )
        )
    except Exception as error:  # noqa: BLE001
        raise IssueUpdateError(str(error)) from error
    updated_issue = result.issue
    if lookup.issue_path.parent == lookup.project_dir / "issues":
        event_id = events[0].event_id if events else None
        publish_issue_mutation(
            root,
            lookup.project_dir,
            updated_issue,
            event_id,
            "issue.mutated",
        )
    return IssueUpdateResult(issue=updated_issue, changed=True)


def _find_duplicate_title(
    issues_dir: Path, title: str, current_identifier: str
) -> Optional[str]:
    normalized_title = title.strip().casefold()
    for issue_path in issues_dir.glob("*.json"):
        if issue_path.stem == current_identifier:
            continue
        try:
            issue = read_issue_from_file(issue_path)
        except (ValueError, ValidationError):
            continue
        if issue.title.strip().casefold() == normalized_title:
            return issue.identifier
    return None
